rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -0,0 +1,114 @@
+        to use signed integers to detect the amounts.
+        """
+        out_event = in_event = None
+        if tx_log.topics[0] == SWAP_SIGNATURE:
+            received_eth = ZERO
+            for event in decoded_events:
+                if event.asset == A_ETH and event.event_type == HistoryEventType.RECEIVE:
+                    received_eth += event.balance.amount
+
+            buyer = hex_or_bytes_to_address(tx_log.topics[2])
+            # Uniswap V3 represents the delta of tokens in the pool with a signed integer
+            # for each token. In the transaction we have the difference of tokens in the pool
+            # for the token0 [0:32] and the token1 [32:64]. If that difference is negative it
+            # means that the tokens are leaving the pool (the user receives them) and if it is
+            # positive they get into the pool (the user sends them to the pool)
+            delta_token_0 = hex_or_bytes_to_int(tx_log.data[0:32], signed=True)
+            delta_token_1 = hex_or_bytes_to_int(tx_log.data[32:64], signed=True)
+            if delta_token_0 > 0:
+                amount_received, amount_sent = delta_token_1, delta_token_0
+            else:
+                amount_received, amount_sent = delta_token_0, delta_token_1
+            # We take the absolute value since we have it negative and we need it positive
+            amount_received = abs(amount_received)
+
+            for event in decoded_events:
+                # When swapping token for ETH the WETH contract is called by the router and the
+                # swap is not executed with the user in the topic but the router. This is when
+                # tx_log.topics[1] == tx_log.topics[2]
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    (event.location_label == buyer or tx_log.topics[1] == tx_log.topics[2]) and
+                    (
+                        event.balance.amount == (spent_amount := asset_normalized_value(amount=amount_sent, asset=event.asset)) or  # noqa: E501
+                        event.asset == A_ETH and spent_amount + received_eth == event.balance.amount  # noqa: E501
+                    )
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.SPEND
+                    event.counterparty = CPT_UNISWAP_V3
+                    event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in uniswap-v3 from {event.location_label}'  # noqa: E501
+                    out_event = event
+                elif (
+                    (
+                        (event.event_type in (HistoryEventType.RECEIVE, HistoryEventType.TRANSFER)) and  # noqa: E501
+                        (event.location_label == buyer or event.asset == A_ETH) and
+                        event.balance.amount == asset_normalized_value(amount=amount_received, asset=event.asset)  # noqa: E501
+                    )
```

Comment:
unnecessary parentheses

```suggestion
                        (event.event_type in (HistoryEventType.RECEIVE, HistoryEventType.TRANSFER)) and  # noqa: E501
                        (event.location_label == buyer or event.asset == A_ETH) and
                        event.balance.amount == asset_normalized_value(amount=amount_received, asset=event.asset)  # noqa: E501
```